lib/Sema/CSBindings.cpp

```diff
@@ -204,14 +204,26 @@ ConstraintSystem::getPotentialBindings(TypeVariableType *typeVar) {
 
     case ConstraintKind::BridgingConversion:
     case ConstraintKind::CheckedCast:
-    case ConstraintKind::DynamicTypeOf:
     case ConstraintKind::EscapableFunctionOf:
     case ConstraintKind::OpenedExistentialOf:
     case ConstraintKind::KeyPath:
     case ConstraintKind::KeyPathApplication:
       // Constraints from which we can't do anything.
       continue;
 
+    case ConstraintKind::DynamicTypeOf: {
+      // Direct binding of the left-hand side could result
+      // in `DynamicTypeOf` failure if right-hand side is
+      // bound (because 'Bind' requires equal types to
+      // succeed), or left is bound to Any which is not an
+      // [existential] metatype.
+      if (constraint->getFirstType()->isEqual(typeVar))
+        return result;
```

Comment:
Returning here is definitely the wrong thing to do. Consider that you might have already processed other constraints involving this type variable and added bindings. You're now returning, which allows those potential bindings to be tested.

I think the right thing to do here instead is to either set `result.InvolvesTypeVariables` any time you're gathering bindings for the `LHS` and the simplified `RHS` is a type variable, or introduce a new flag, similar to what I started working on for `BindParam` in https://github.com/apple/swift/pull/11215.